src/Features/Core/Portable/AddImport/CodeActions/AssemblyReferenceCodeAction.cs

```diff
@@ -16,47 +17,48 @@ internal abstract partial class AbstractAddImportFeatureService<TSimpleNameSynta
             {
                 Contract.ThrowIfFalse(fixData.Kind == AddImportFixKind.ReferenceAssemblySymbol);
-                _lazyResolvedPath = new Lazy<string>(ResolvePath);
             }
 
-            private string ResolvePath()
+            private Task<string?> ResolvePathAsync(CancellationToken cancellationToken)
             {
-                var assemblyResolverService = OriginalDocument.Project.Solution.Workspace.Services.GetService<IFrameworkAssemblyPathResolver>();
+                var assemblyResolverService = OriginalDocument.Project.Solution.Workspace.Services.GetRequiredService<IFrameworkAssemblyPathResolver>();
 
-                var assemblyPath = assemblyResolverService?.ResolveAssemblyPath(
+                return assemblyResolverService.ResolveAssemblyPathAsync(
                     OriginalDocument.Project.Id,
                     FixData.AssemblyReferenceAssemblyName,
-                    FixData.AssemblyReferenceFullyQualifiedTypeName);
-
-                return assemblyPath;
+                    FixData.AssemblyReferenceFullyQualifiedTypeName,
+                    cancellationToken);
             }
 
-            internal override bool PerformFinalApplicabilityCheck
-                => true;
+            protected override Task<IEnumerable<CodeActionOperation>> ComputePreviewOperationsAsync(CancellationToken cancellationToken)
+                => ComputeOperationsAsync(isPreview: true, cancellationToken);
 
-            internal override bool IsApplicable(Workspace workspace)
-                => !string.IsNullOrWhiteSpace(_lazyResolvedPath.Value);
+            protected override Task<IEnumerable<CodeActionOperation>> ComputeOperationsAsync(CancellationToken cancellationToken)
+                => ComputeOperationsAsync(isPreview: false, cancellationToken);
 
-            protected override async Task<IEnumerable<CodeActionOperation>> ComputeOperationsAsync(CancellationToken cancellationToken)
+            private async Task<IEnumerable<CodeActionOperation>> ComputeOperationsAsync(bool isPreview, CancellationToken cancellationToken)
             {
-                var service = OriginalDocument.Project.Solution.Workspace.Services.GetService<IMetadataService>();
-                var resolvedPath = _lazyResolvedPath.Value;
-                var reference = service.GetReference(resolvedPath, MetadataReferenceProperties.Assembly);
-
                 var newDocument = await GetUpdatedDocumentAsync(cancellationToken).ConfigureAwait(false);
+                var newProject = newDocument.Project;
 
                 // Now add the actual assembly reference.
-                var newProject = newDocument.Project;
-                newProject = newProject.WithMetadataReferences(
-                    newProject.MetadataReferences.Concat(reference));
+                if (!isPreview)
+                {
+                    var resolvedPath = await ResolvePathAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
so ComputeOperationsAsync now switches to the UI thread internally since ResolveAssemblyPathAsync switches.  So this removes the deadlock risk in the LSP CodeActionsHandler since it just retrieves and filters, but it remains in the CodeActionResolveHandler (calls GetOperationsAsync) and RunCodeActionsHandler (applies the operations).  Is that what's happening?

Reply:
Asynchronous operations are generally allowed to switch to the UI thread. Roslyn deviates from these rules, but not on the code fix _application_ path.